enhancements/console/dynamic-plugins.md

```diff
@@ -0,0 +1,372 @@
+
+In addition to static plugins, we'd like to introduce dynamic plugins. Dynamic
+plugins are contributed by operators and loaded at runtime. Eventually we want
+to migrate static plugins to dynamic plugins to decouple them from the console
+and allow plugins to be updated more frequently.
+
+## Motivation
+
+Static plugins have worked well, but they have some limitations:
+
+1. Static plugins can only be updated on the OpenShift release cadence. We want
+   to enable plugin teams to move faster.
+1. Static plugins are tied to the console version, not the component operator
+   version. We want to be able to update plugins as operators change APIs or
+   add features. The right version of the plugin should be used for the
+   installed version of the operator.
+
+We already have a half-dozen teams contributing to console. We need to be able
+to scale as more teams contribute. This will be difficult if everyone is
+contributing to a single monorepo.
+
+### Goals
+
+* Operators can deliver console plugins separate from the console image and
+  update plugins when the operator updates.
+* The dynamic plugin API is similar to the static plugin API to ease migration.
+* Plugins can use shared console components such as list and details page components.
+* Shared components from core will be part of a well-defined plugin API.
+* Plugins can use [Patternfly 4](https://www.patternfly.org/v4/) components.
+* Cluster admins control what plugins are enabled.
+* Misbehaving plugins should not break console.
+* Existing static plugins are not affected and will continue to work as expected.
+
+### Non-Goals
+
+* Initially we don't plan to make this a public API. The target use is for Red
+  Hat operators. We might reevaluate later when dynamic plugins are more
+  mature.
+* We can't prevent breaking changes in Patternfly APIs the console exposes to plugins.
+* Plugins won't be sandboxed. They will have full JavaScript access to the DOM and network.
+* This proposal does not cover allowing plugins to contribute backend console endpoints.
+
+## Proposal
+
+### Module Federation
+
+Console will use [Webpack module federation](https://webpack.js.org/concepts/module-federation/)
+to load plugins at runtime. Module federation allows a JavaScript application
+to dynamically load code from another application while sharing dependencies.
+It also allows console to share its components with plugins built and bundled
+separately.
+
+Plugins will need to be built with Webpack 5 to use module federation.
+
+We have a pull request with a [working prototype](https://github.com/openshift/console/pull/6101)
+that uses module federation.
+
+### Shared Dependencies
+
+Console will expose React and Patternfly 4 as shared dependencies to plugins.
+Plugins can use any Patternfly component. Only a single version of React and
+Patternfly will be loaded.
+
+Plugins will not be able to use legacy Patternfly 3 components.
+
+### Delivering Plugins
```

Comment:
Should this be the responsibility of the individual operators, or should we make this a first-class extension of bundles and OLM? 

I am worried that each team will have to write the same code to handle this, and wondering if there is a way to make this simpler.

Reply:
can probably provide a common base image(with the http server) that teams layer their content on top of or maybe just a template Dockerfile (to avoid having to even own/publish the base image).  and sample resource yamls for the service+deployment definition.  

For now w/ the limited number of operators that are doing this, i don't think we need to bend over backwards to make this easy (it's not that hard to do yourself).

Reply:
They may not even want to host this content separately from other servers they are already running, why run another binary if it is not necessary. I think it would be a mistake to tie this to operators. If we have any plan in the future to allow customers to use this as a way to extend the console experience, I don't think having an operator is a pre-req to that.

Reply:
To clarify my statement, I think for us, shipping this content by delivering it as part of a products overall operator deployment is the correct approach. But the ConsolePlugin API itself doesn't necessitate an operator is managing the service providing the assets.

Reply:
yeah, agreed.  So i would categorize doing "nice things for operators to make their lives easy" here as entirely optional/nice-to-have and not a requirement of the EP implementation.